### PR TITLE
Add tracing option

### PR DESCRIPTION
**Pull request by johnbelamaric**
This adds an option to enable tracing with OpenTracing/OpenZipkin.
Enabling tracing will have a large impact on performance so it is
not advisable in production.

### CODE REVIEW

**Comment by johnbelamaric** [comment]
@miekg let me know what you think of this. I didn't see any real impact to performance when it is disabled (though my testing was only with proxy module).

To try it out run Zipkin in docker:
```
docker run -d -p 9411:9411 openzipkin/zipkin
```
Then when you run CoreDNS:

```
./coredns -conf Corefile -dns.trace zipkinhost:9411
```

Then you can navigate to the browser at zipkinhost:9411 to see the trace results.

**Comment by miekg** [comment]
Ack. Currently AFK, but I assume you can just flip this on and off and
reload coredns?

Sounds useful although I haven't played with opentrace

**Comment by johnbelamaric** [comment]
Yes, it is enabled with the `-dns.trace` option.

**Comment by miekg** [comment]
I think for hot reloading it needs to be in the corefile?

**Comment by johnbelamaric** [comment]
Ah, ok. Didn’t realize there was hot reloading. It’s not a middleware, and I didn’t see a way to do global settings in the Corefile. Right now it’s global but it could actually be per server - but two stanzas in the Corefile may apply to the same server so not sure how to handle that.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/coredns] Add tracing opt..." ]
>Ah, ok. Didn’t realize there was hot reloading. It’s not a middleware, and I didn’t see a way to do global settings in the Corefile. Right now it’s global but it could actually be per server - but two stanzas in the Corefile may apply to the same server so not sure how to handle that.
This could just be an option in the config of a server? Just like the *root*
directive?

**Comment by johnbelamaric** [comment]
I could do that. But doesn't it gather all the zones on the same port into one server? So what would I do with this:

```
foo.com:53 {
  trace 1.2.3.4:9411
  log stdout
  errors
  proxy foo.com /etc/resolv.conf
}

foobar.com:53 {
  trace 9.9.9.9:12345
  log stdout
}
```

Don't those end up being service by the same dnsserver.Server?

I guess I can stick the tracer on the Config instead of the server, that would allow this to work as one would expect.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/coredns] Add tracing opt..." ]
>I could do that. But doesn't it gather all the zones on the same port into one server? So what would I do with this:
>
>```
>foo.com:53 {
>  trace 1.2.3.4:9411
>  log stdout
>  errors
>  proxy foo.com /etc/resolv.conf
>}
>
>foobar.com:53 {
>  trace 9.9.9.9:12345
>  log stdout
>}
>```
>
>Don't those end up being service by the same dnsserver.Server?
>
>I guess I can stick the tracer on the Config instead of the server, that would allow this to work as one would expect.

Yes

**Comment by miekg** [comment]
Think this should be a middleware just like metrics (functionality is simular-ish anyway). Might need some bits in core, but those can be minimal, i guesss

**Comment by johnbelamaric** [comment]
Ok

**Comment by johnbelamaric** [comment]
Ok, it's a middleware now, that works fine. Also I put in the skeleton to allow other OpenTracing-compatible implementations other than Zipkin.

File: middleware/middleware.go [hunk 1]
```diff
@@ -70,6 +71,12 @@ func Error(name string, err error) error { return fmt.Errorf("%s/%s: %s", "middl
 // and a nil error.
 func NextOrFailure(name string, next Handler, ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
 	if next != nil {
+		span := ot.SpanFromContext(ctx)
+		if span != nil {
```
Comment (miekg) [inline middleware/middleware.go]: Nit: if statement can be done on 1 line.
`if span := ot.SpanFromContext(ctx); span != nil {`

File: middleware/trace/README.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
+# trace
+
+This module enables OpenTracing-based tracing of DNS requests as they go through the
+middleware chain.
+
+## Syntax
+
+~~~
+trace [endpoint-type] [endpoint]
```
Comment (miekg) [inline middleware/trace/README.md]: please uppercase endpoint-type and endpoint to be inline with the other READMEs

File: middleware/trace/README.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+~~~
+
+For each server you which to trace.
+
+It optionally the endpoint type and endpoint. The type defaults to Zipkin and the
```
Comment (miekg) [inline middleware/trace/README.md]: `**ENDPOINT-TYPE**` and `**ENDPOINT**`

**Comment by miekg** [comment]
lgtm. Please update; https://github.com/miekg/coredns/wiki/CoreDNS-005-Release
Maybe also add some works about the k8s middleware there?

**Comment by johnbelamaric** [comment]
Ok, edits to release notes also done.

**Review by miekg — APPROVED** [review]

**Pull request merged**